src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/Inflater.cs

```diff
@@ -98,6 +107,47 @@ public unsafe int InflateVerified(byte* bufPtr, int length)
             }
         }
 
+        /// <summary>
+        /// If this stream has some input leftover that hasn't been processed then we should
+        /// check if it is another GZip file concatenated with this one.
+        /// 
+        /// Returns false if the leftover input is another GZip data stream.
+        /// </summary>
+        private unsafe bool ResetStreamForLeftoverInput()
+        {
+            Debug.Assert(!NeedsInput());
+            Debug.Assert(IsGzipStream());
+            Debug.Assert(_inputBufferHandle.IsAllocated);
+
+            lock (SyncLock)
+            {
+                IntPtr nextInPtr = _zlibStream.NextIn;
+                byte* nextInPointer = (byte*)nextInPtr.ToPointer();
+                uint nextAvailIn = _zlibStream.AvailIn;
+
+                // Check the leftover bytes to see if they start with he gzip header ID bytes
+                if (*nextInPointer != ZLibNative.GZip_Header_ID1 || (nextAvailIn > 1 && *(nextInPointer + 1) != ZLibNative.GZip_Header_ID2))
```

Comment:
How do we know at this point that nextInPointer is safe to dereference?

Reply:
This function is only called in one place, and only after a check that the inputbufferhandle is allocated. the nextIn pointer is always set when inputbufferhandle is allocated, and it always points to a byte within the allocated section. the zlib native code moves the nextin pointer during its operation, but only as long as `availin` is greater than 0. `availin` is set whenever the inputbufferhandle is allocated, and is set to the length of the buffer pointed to by the handle.

Reply:
Should we be asserting then that nextAvailIn is > 0?

Reply:
sure.